### PR TITLE
Improve handling of various lifecycle erorrs in VNet

### PR DESCRIPTION
**Pull request by ravicious**
This PR also prepares VNet for the integration with Connect. I described reasoning behind each change in commit messages.

A key piece of information to understand about VNet is that in this version we start a separate tsh command, `tsh vnet-admin-setup`. For now, we use `osascript -e "do shell script … with administrator privileges"` to execute this command with root privileges. This gives us the ability to create TUN devices and make system-wide DNS changes.

This command is started with [`os/exec.CommandContext`](https://pkg.go.dev/os/exec#CommandContext). The semantics of running a command through osascript are a little bit different, but ultimately they don't impact the implementation too much. E.g., cancelling the context passed to `os/exec.CommandContext` kills the osascript invocation, but not exactly `tsh vnet-admin-setup` since an unprivileged process can't kill privileged one. The admin subcommand quits when it detects that the socket, which is used to pass the TUN device, has been closed.

---

When it comes to testing the admin subcommand quitting prematurely, the way I do is I start `tsh vnet` and then I open Activity Monitor and search for `tsh`. If you go to View -> All Processes, Hierarchically, the admin subcommand is going to be under kernel_task -> launchd -> authtrampoline -> bash -> tsh (click on an item while pressing Option to recursively expand items). From there, you can get the PID to kill it (`sudo kill -s KILL <pid>`).

If you want to cause the admin subcommand to quit because of the removed socket, `tsh vnet -d` is going to report the socket path in the second log message. You don't need sudo to remove the socket.

Cases I tested:

* Interrupting `tsh vnet` while osascript prompt is shown.
* Interrupting `tsh vnet` after going through the osascript prompt (how the user would usually interrupt it).
* Removing the socket file while `tsh vnet -d` is running.
* Killing the admin process while `tsh vnet` is running.

---

We should also use `lib/utils/log.NewPackageLogger` instead of calling functions directly on `slog`. But I don't want to introduce too many conflicts right now since we have a couple of other VNet PRs open at the moment.

### CODE REVIEW

File: lib/vnet/setup_darwin.go
Comment (nklaassen) [inline lib/vnet/setup_darwin.go:73]: is there a reason this needs its own context and you can't use the same context as everything else? can't we do a `ctx, cancel := context.WithCancel(ctx)` and use the same context for everything? i think the fewer different contexts we have to keep track of the better

You may not need the WithCancel at all, since if any of the errgroup functions returns an error, the errgroup will do the cancellation. Make your own WithCancel context if we need to worry about execAdminSubcommand returning nil prematurely

File: lib/vnet/setup.go
Comment (nklaassen) [inline lib/vnet/setup.go:109]: note: i think i need to improve/change the socket based keepalive for the admin process, i noticed it never exits if the non-root process gets killed without getting a chance to close/delete the socket, like when i run with a debugger. probably need to go back to the PID approach

Comment (ravicious) [reply lib/vnet/setup.go:109]: The socket approach was so nice from Connect's perspective though, because I could just shut down the socket and the admin process would shut down too. With the PID approach, the only thing that comes to my mind is hacks like removing the PID file to make the admin process think that the unprivileged process has quit.

Comment (nklaassen) [reply lib/vnet/setup.go:109]: Oh right, yeah let's avoid any PID approach. Another idea is to keep the user process listening on the socket and have the root process make sure it can continually dial to it, that way if the user process dies without deleting the socket, the dial will fail, and the root process can clean up

Comment (nklaassen) [reply lib/vnet/setup.go:109]: Gah that would be bad when running with a debugger too though, if the user process doesn't get a chance to accept a connection from the socket often enough

File: lib/vnet/setup_darwin.go
Comment (ravicious) [reply lib/vnet/setup_darwin.go:216]: Right, I think what you proposed here is better. Though in general using a separate goroutine + a buffered err channel is not a bad way to make blocking calls "context-aware", is it?

I've already run into this a couple of times in Go where I have to make a blocking (sys)call but at the same time I'd like to make the surrounding code not be blocked on it if the context gets canceled. I don't know what's the usual way to tackle this.

Here the main problem with this approach was that cancellation of the context itself affects the blocking call, causing a race condition.

File: lib/vnet/setup.go
Comment (ravicious) [reply lib/vnet/setup.go:109]: Did you already experiment with passing an extra file descriptor to osascript and then weaving it through to the admin subcommand? https://stackoverflow.com/a/18732806/742872

This way we could communicate through a pipe with the admin subcommand. OTOH the example above doesn't use "with administrator privileges", I guess osascript probably doesn't forward fds when it does that.

File: lib/vnet/setup_darwin.go
Comment (ravicious) [reply lib/vnet/setup_darwin.go:73]: Right, we should be able to use a single context.

The reason I added a separate one was that I wanted to make sure the socket is closed after the admin cmd exits prematurely, but the check for a premature exit is done at a completely different place than the goroutine that actually runs the errgroup. So in the previous version, the goroutine executing the admin subcommand would exit with nil and the socket would be kept open.

If I move the check for a premature exit inside the same goroutine that executes the command, then just a single context should be enough.

**Comment by nklaassen** [comment]
Since I'm still making edits to `nklaassen/vnet5` and I've already rebased a few times and I'll probably continue to, I'd recommend making your own branch at wherever you originally branched this off `nklaassen/vnet5`, and using that as the base branch for this PR, so you have a stable diff here. You can sync that branch to `nklaassen/vnet5`on your own schedule, and once it merges you can rebase this to master

**Comment by ravicious** [comment]
I'll wait until we settle on https://github.com/gravitational/teleport/pull/41653, then I'll split that PR into `lib/vnet` and `tool/tsh` changes that will go into this PR and `lib/teleterm` changes that will go into https://github.com/gravitational/teleport/pull/41415.

File: lib/vnet/setup.go
Comment (nklaassen) [reply lib/vnet/setup.go:109]: Ultimately I think we should do a PID file at a static path `/var/run/vnet.pid`. This can prevent a user from running multiple VNet processes at once, and also solve the admin process termination issue

We should only create it when starting VNet and delete it when stopping, Connect should be able to handle that, it won't necessarily be tied to the full process lifetime

When tsh or Connect try to start VNet, they should check if this file exists. If not, they should create it with their own PID.

if the file does exist, check if the process with the PID is actually running. If not, assume it died without cleaning up the file, and overwrite it. If that process is running, return an error and fail to start vnet since it is already running in another process

the admin process should watch that file and terminate when either: it is deleted or the PID in it changes or the process associated with the PID in the file is not running.

I'm not sure where to make this change at this point to reduce conflicts with you. Maybe after you get these current lifecycle improvement PRs sorted out, I can base it off your newest branch

**Comment by ravicious** [comment]
@zmb3 For the most part the code here is identical to what you already saw in #41653. The first couple of commits include a messy implementation that was cleaned up by the final commit, which is a squashed version of #41653.

File: tool/tsh/common/vnet_darwin.go
Comment (nklaassen) [inline tool/tsh/common/vnet_darwin.go:68]: yeah again I think it's fine to return the error even if we got killed by ctrl-c. If you kill any other long-running command like `tsh login` (waiting for password/mfa) it returns an error and that's fine

this also avoids just returning nil if we some returned context.Canceled without the user actually trying to kill vnet with ctrl-c

```suggestion
	return trace.Wrap(processManager.Wait())
```

File: lib/vnet/setup.go
Comment (nklaassen) [inline lib/vnet/setup.go:68]: since this is tied to the process lifetime, i think we might as well make it a critical task on the process manager, and that gives us a guarantee that something will wait for the goroutine to terminate and things will fail loudly if this somehow exits early

```suggestion
	pm.AddCriticalBackgroundTask("socket closer", func() error {
		// Keep the socket open until the process context is canceled.
		// Closing the socket signals the admin subcommand to terminate.
		<-processCtx.Done()
		return trace.Wrap(socket.Close())
	})
```

**Review by nklaassen — APPROVED** [review]

File: tool/tsh/common/vnet_darwin.go
Comment (ravicious) [reply tool/tsh/common/vnet_darwin.go:68]: I think you're generally right and I was going to bring up `tsh login` as an example too – it just got stuck at reading my MFA key and Ctrl+C returned context canceled + a useful user message as to where the code got stuck.

I'm trying to build a good mental model for this. For example, `tsh proxy app` emits no message on context canceled after the proxy is up and returns 0 exit code. `tail` emits no message on Ctrl+C, but the exit code is 130 (command exited through sigint).

Originally, I added checks for context.Canceled because it was annoying to see "context canceled" when I wanted to close VNet, it seemed like a bad UX.

Now I realize that maybe how we depend on ctx and how we always print the error message is wrong? That is, VNet should somehow wait for SIGINT and if it's received while VNet is idle, "context canceled" should not be presented to the user. It should be emitted though while VNet is waiting for something to happen.

For now I'll just always return the error.

File: lib/vnet/setup_test.go
Comment (ravicious) [reply lib/vnet/setup_test.go:60]: Hmm I think the current version is fine. `taskErr` should really have been called `expectedErr`. I want to check that `err` includes `expectedErr`, and that the error returned from `pm.Wait` is the same as the one returned from `context.Cause(pmCtx)`.

For now I'll rename `taskErr` to `expectedErr`.

File: lib/vnet/setup.go
Comment (ravicious) [reply lib/vnet/setup.go:68]: That works, but we have to do add an error aggregate:

```go
	pm.AddCriticalBackgroundTask("socket closer", func() error {
		// Keep the socket open until the process context is canceled.
		// Closing the socket signals the admin subcommand to terminate.
		<-processCtx.Done()
		return trace.NewAggregate(processCtx.Err(), socket.Close())
	})
```

This makes the task behave like other context-aware functions (as in, they return `context.Canceled` when idle when the context gets canceled). Otherwise after Ctrl+C, this would race with the other tasks – if `socket.Close()` returned with no error, the user would see `critical task "socket closer" exited prematurely`.

That's because `errgroup.Group.Wait` does not await the context by itself, it waits for the first goroutine to finish. And `processManager.Wait` lost the ability to check if the context was canceled after we moved the context outside of the struct.

I could also do something like this, I'm not sure if this is better:

```go
	pm.AddCriticalBackgroundTask("socket closer", func() error {
		// Keep the socket open until the process context is canceled.
		// Closing the socket signals the admin subcommand to terminate.
		<-processCtx.Done()
		err := processCtx.Err()
		if closeErr := socket.Close(); closeErr != nil {
			err = closeErr
		}
		return trace.Wrap(err)
	})
```

This socket closer is essentially a cleanup function, that's why I also felt like it shouldn't be deemed a critical task.

For now I'm going to turn it into a critical task and add an aggregate.

Comment (ravicious) [reply lib/vnet/setup.go:68]: https://github.com/golang/go/issues/23595#issuecomment-361633753

> The purpose of the `errgroup` package is to coordinate error collection with cancellation, since that's fairly subtle to get right using `sync.WaitGroup`.
If you cancel the `Context` on the first error, the subsequent errors tend not to be meaningful (they're often either `context.Canceled` or other errors that arise as a side-effect of cancellation).
> 
> If you want all of the errors (and don't want to cancel the `Context`), then you may as well just use a `sync.WaitGroup`. It's not a lot of boilerplate, and not particularly error-prone.

File: lib/vnet/vnet.go
Comment (zmb3) [inline lib/vnet/vnet.go:151]: :+1: for the better name here!

File: lib/vnet/vnet.go
Comment (zmb3) [inline lib/vnet/vnet.go:186]: nit: the mutex should be considered an implementation detail. If we change it in the future I doubt we will remember to find and update this comment.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**
